hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/ObservableResultsTest.java

```diff
@@ -0,0 +1,240 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+public class ObservableResultsTest extends TestInClusterSupport {
+
+    private String observableName;
+    private TestObserver testObserver;
+    private Observable<Long> testObservable;
+
+    @Before
+    public void before() {
+        observableName = randomName();
+        testObserver = new TestObserver();
+        testObservable = jet().getObservable(observableName);
+        testObservable.addObserver(testObserver);
+    }
+
+    @After
+    public void after() {
+        testObservable.destroy();
+    }
+
+    @Test
+    public void batchJobCompletesSuccessfully() {
+        Pipeline pipeline = Pipeline.create();
+        pipeline.readFrom(TestSources.items(0L, 1L, 2L, 3L, 4L))
+                .writeTo(Sinks.observable(observableName));
+
+        Job job = jet().newJob(pipeline);
+        job.join();
+
+        assertSortedValues(testObserver, 0L, 1L, 2L, 3L, 4L);
+        assertError(testObserver, null);
+        assertCompletions(testObserver, 1);
+    }
+
+    @Test
+    public void batchJobFails() {
+        BatchSource<String> errorSource = SourceBuilder
+                .batch("error-source", x -> (Object) null)
+                .<String>fillBufferFn((in, Void) -> {
+                    throw new Exception("Ooops!");
+                })
+                .destroyFn(ConsumerEx.noop())
+                .build();
+
+        Pipeline pipeline = Pipeline.create();
+        pipeline.readFrom(errorSource)
+                .writeTo(Sinks.observable(observableName));
+
+        Job job = jet().newJob(pipeline);
+        assertTrueEventually(() -> assertEquals(JobStatus.FAILED, job.getStatus()));
+
+        assertSortedValues(testObserver);
+        assertError(testObserver, "Ooops!");
+        assertCompletions(testObserver, 0);
+    }
+
+    @Test
+    public void streamJob() {
+        Pipeline pipeline = Pipeline.create();
+        pipeline.readFrom(TestSources.itemStream(100))
+                .withoutTimestamps()
+                .map(SimpleEvent::sequence)
+                .writeTo(Sinks.observable(observableName));
+
+        Job job = jet().newJob(pipeline);
+
+        assertTrueEventually(() -> assertTrue(testObserver.getSortedValues().size() > 10));
+        assertError(testObserver, null);
+        assertCompletions(testObserver, 0);
+
+        job.cancel();
+        assertTrueEventually(() -> assertTrue(testObserver.getSortedValues().size() > 10));
+        assertError(testObserver, null);
```

Comment:
This is not correct. `java.util.concurrent.CancellationException` is thrown during `job.cancel()` - this assert passes only before it is actually faster then Exception observing. If some sleep is added before assert then it fails.